### PR TITLE
Create configuration_script_payloads API doc

### PR DESCRIPTION
**Pull request by rwellon**
This is a draft for creating the API doc for the /api/configuration_script_payloads

### CODE REVIEW

**Comment by rwellon** [comment]
@miq-bot assign @agrare
@miq-bot add_reviewer @Fryguy
@miq-bot add-label quinteros/yes?

File: api/reference/configuration_script_payloads.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+### Querying Configuration Script Payloads
+
+Query all configuration script payloads in the system:
+
+    GET /api/configuration_script_payloads
+
+Get the details on a specific configuration script payload:
+
+    GET /api/configuration_script_payloads/:id
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:33]:
```json
{
  "name": "configuration_script_payloads",
  "count": 2,
  "subcount": 2,
  "pages": 1,
  "resources": [
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1"
    },
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1"
    }
  ],
  "actions": [
    {
      "name": "edit",
      "method": "post",
      "href": "http://localhost:3000/api/configuration_script_payloads"
    }
  ],
  "links": {
    "self": "http://localhost:3000/api/configuration_script_payloads?offset=0",
    "first": "http://localhost:3000/api/configuration_script_payloads?offset=0",
    "last": "http://localhost:3000/api/configuration_script_payloads?offset=0"
  }
}

File: api/reference/configuration_script_payloads.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+- [Querying Configuration Script Payload Authentication](#querying-configuration-script-payload-authentication)
+
+- [Creating Configuration Script Payload Authentication](#creating-configuration-script-payload-authentication)
+
+### Querying Configuration Script Payloads
+
+Query all configuration script payloads in the system:
+
+    GET /api/configuration_script_payloads
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:29]:
```json
{
  "name": "configuration_script_payloads",
  "count": 2,
  "subcount": 2,
  "pages": 1,
  "resources": [
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1"
    },
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1"
    }
  ],
  "actions": [
    {
      "name": "edit",
      "method": "post",
      "href": "http://localhost:3000/api/configuration_script_payloads"
    }
  ],
  "links": {
    "self": "http://localhost:3000/api/configuration_script_payloads?offset=0",
    "first": "http://localhost:3000/api/configuration_script_payloads?offset=0",
    "last": "http://localhost:3000/api/configuration_script_payloads?offset=0"
  }
}

File: api/reference/configuration_script_payloads.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+Patch method:
+
+    PATCH /api/configuration_script_payloads/:id
+
+### Querying Configuration Script Payload Authentication
+
+Query the authentication credentials for a specific configuration script payload:
+
+    GET /api/configuration_script_payloads/:id/authentications
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:57]:
```json
{
  "name": "authentications",
  "count": 5,
  "subcount": 0,
  "pages": 1,
  "resources": [],
  "actions": [
    {
      "name": "create",
      "method": "post",
      "href": "http://localhost:3000/api/configuration_script_payloads/1/authentications"
    }
  ],
  "links": {
    "self": "http://localhost:3000/api/configuration_script_payloads/1/authentications?offset=0",
    "first": "http://localhost:3000/api/configuration_script_payloads/1/authentications?offset=0",
    "last": "http://localhost:3000/api/configuration_script_payloads/1/authentications?offset=0"
  }
}
```

File: api/reference/configuration_script_payloads.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Get the details on a specific authentication credential for a configuration script payload:
+
+    GET /api/configuration_script_payloads/:id/authentications/:id
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:61]:
```json
{
  "href": "http://localhost:3000/api/configuration_script_payloads/2/authentications/3",
  "id": "3",
  "name": "manageiq_api",
  "authtype": null,
  "userid": "admin",
  "password": null,
  "resource_id": "25",
  "resource_type": "ExtManagementSystem",
  "created_on": "2023-11-15T19:59:23Z",
  "updated_on": "2023-11-15T19:59:23Z",
  "last_valid_on": null,
  "last_invalid_on": null,
  "credentials_changed_on": null,
  "status": null,
  "status_details": null,
  "type": "ManageIQ::Providers::Workflows::AutomationManager::WorkflowCredential",
  "auth_key": null,
  "fingerprint": null,
  "service_account": null,
  "public_key": null,
  "manager_ref": "3",
  "options": null,
  "evm_owner_id": null,
  "miq_group_id": "1",
  "tenant_id": "1",
  "become_username": null,
  "become_password": null,
  "auth_key_password": null,
  "ems_ref": "manageiq_api",
  "become_method": null
}
```

File: api/reference/configuration_script_payloads.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+Get the details on a specific configuration script payload:
+
+    GET /api/configuration_script_payloads/:id
+
+### Editing configuration script payloads
+
+Configuration script payloads can be edited by using a POST to the configuration script payloads collection:
+
+    POST /api/configuration_script_payloads
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:39]:
```
POST /api/configuration_script_payloads
{
  "action": "edit",
  "resources": [
    {"id": 1, "description": "A Helpful Description"}
  ]
}
```

```json
{
  "results": [
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1",
      "description": "A Helpful Description",
      "credentials": {
        "api_user.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "userid"
        },
        "api_password.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "password"
        }
      },
      "id": "1",
      "manager_id": "2",
      "manager_ref": null,
      "name": "provision-vm-service/list-providers.asl",
      "variables": null,
      "created_at": "2023-11-08T17:43:30Z",
      "updated_at": "2023-11-15T20:14:28Z",
      "survey_spec": null,
      "inventory_root_group_id": null,
      "type": "ManageIQ::Providers::Workflows::AutomationManager::Workflow",
      "parent_id": null,
      "configuration_script_source_id": "2",
      "run_by_userid": null,
      "payload": "{\n  \"Comment\": \"List Providers.\",\n  \"StartAt\": \"ListProviders\",\n  \"States\": {\n    \"ListProviders\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"docker://docker.io/agrare/list-providers:latest\",\n      \"End\": true,\n      \"Credentials\": {\n        \"api_user.$\": \"$.api_user\",\n        \"api_password.$\": \"$.api_password\",\n        \"api_token.$\": \"$.api_token\",\n        \"api_bearer_token.$\": \"$.api_bearer_token\"\n      },\n      \"Parameters\": {\n        \"API_URL.$\": \"$$.Execution._manageiq_api_url\",\n        \"VERIFY_SSL.$\": \"$.dialog.dialog_verify_ssl\",\n\n        \"PROVIDER_TYPE\": \"ManageIQ::Providers::Vmware::InfraManager\"\n      }\n    }\n  }\n}\n",
      "payload_type": "json",
      "context": null,
      "output": null,
      "status": null,
      "miq_task_id": null,
      "payload_valid": true,
      "payload_error": null
    }
  ]
}
```

File: api/reference/configuration_script_payloads.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+### Updating configuration script payloads
+
+Configuration script payloads can be updated by using a PUT method or PATCH method.
+
+PUT method:
+
+    PUT /api/configuration_script_payloads/:id
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:47]: `PUT /api/configuration_script_payloads/1 {"description": "A Helpful Description"}`

```json
{
  "results": [
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1",
      "description": "A Helpful Description",
      "credentials": {
        "api_user.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "userid"
        },
        "api_password.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "password"
        }
      },
      "id": "1",
      "manager_id": "2",
      "manager_ref": null,
      "name": "provision-vm-service/list-providers.asl",
      "variables": null,
      "created_at": "2023-11-08T17:43:30Z",
      "updated_at": "2023-11-15T20:14:28Z",
      "survey_spec": null,
      "inventory_root_group_id": null,
      "type": "ManageIQ::Providers::Workflows::AutomationManager::Workflow",
      "parent_id": null,
      "configuration_script_source_id": "2",
      "run_by_userid": null,
      "payload": "{\n  \"Comment\": \"List Providers.\",\n  \"StartAt\": \"ListProviders\",\n  \"States\": {\n    \"ListProviders\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"docker://docker.io/agrare/list-providers:latest\",\n      \"End\": true,\n      \"Credentials\": {\n        \"api_user.$\": \"$.api_user\",\n        \"api_password.$\": \"$.api_password\",\n        \"api_token.$\": \"$.api_token\",\n        \"api_bearer_token.$\": \"$.api_bearer_token\"\n      },\n      \"Parameters\": {\n        \"API_URL.$\": \"$$.Execution._manageiq_api_url\",\n        \"VERIFY_SSL.$\": \"$.dialog.dialog_verify_ssl\",\n\n        \"PROVIDER_TYPE\": \"ManageIQ::Providers::Vmware::InfraManager\"\n      }\n    }\n  }\n}\n",
      "payload_type": "json",
      "context": null,
      "output": null,
      "status": null,
      "miq_task_id": null,
      "payload_valid": true,
      "payload_error": null
    }
  ]
}
```

File: api/reference/configuration_script_payloads.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+Patch method:
+
+    PATCH /api/configuration_script_payloads/:id
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:51]: `PATCH /api/configuration_script_payloads/1 {"description": "A Helpful Description"}`

```json
{
  "results": [
    {
      "href": "http://localhost:3000/api/configuration_script_payloads/1",
      "description": "A Helpful Description",
      "credentials": {
        "api_user.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "userid"
        },
        "api_password.$": {
          "credential_ref": "manageiq_api",
          "credential_field": "password"
        }
      },
      "id": "1",
      "manager_id": "2",
      "manager_ref": null,
      "name": "provision-vm-service/list-providers.asl",
      "variables": null,
      "created_at": "2023-11-08T17:43:30Z",
      "updated_at": "2023-11-15T20:14:28Z",
      "survey_spec": null,
      "inventory_root_group_id": null,
      "type": "ManageIQ::Providers::Workflows::AutomationManager::Workflow",
      "parent_id": null,
      "configuration_script_source_id": "2",
      "run_by_userid": null,
      "payload": "{\n  \"Comment\": \"List Providers.\",\n  \"StartAt\": \"ListProviders\",\n  \"States\": {\n    \"ListProviders\": {\n      \"Type\": \"Task\",\n      \"Resource\": \"docker://docker.io/agrare/list-providers:latest\",\n      \"End\": true,\n      \"Credentials\": {\n        \"api_user.$\": \"$.api_user\",\n        \"api_password.$\": \"$.api_password\",\n        \"api_token.$\": \"$.api_token\",\n        \"api_bearer_token.$\": \"$.api_bearer_token\"\n      },\n      \"Parameters\": {\n        \"API_URL.$\": \"$$.Execution._manageiq_api_url\",\n        \"VERIFY_SSL.$\": \"$.dialog.dialog_verify_ssl\",\n\n        \"PROVIDER_TYPE\": \"ManageIQ::Providers::Vmware::InfraManager\"\n      }\n    }\n  }\n}\n",
      "payload_type": "json",
      "context": null,
      "output": null,
      "status": null,
      "miq_task_id": null,
      "payload_valid": true,
      "payload_error": null
    }
  ]
}
```

File: api/reference/configuration_script_payloads.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Get the details on a specific authentication credential for a configuration script payload:
+
+    GET /api/configuration_script_payloads/:id/authentications/:id
+
+### Creating Configuration Script Payload Authentication
+
+Authentication credentials for a configuration script payload can be created by using a POST to the configuration script payload collection:
+
+    POST /api/configuration_script_payloads/:id/authentications
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:67]: We don't use this for Workflows, still helpful for Embedded Ansible but I'll have to dig in a bit

File: api/reference/configuration_script_payloads.md (hunk 5 above)
Comment (agrare) [reply api/reference/configuration_script_payloads.md:39]: Mapping credentials:
```
POST /api/configuration_script_payloads
{
  "action": "edit",
  "resources": [
    {
      "id": 1,
      "credentials": [
        {"my-cred-user": {"credential_ref": "my-credential", "credential_field": "userid"},
        {"my-cred-password": {"credential_ref": "my-credential", "credential_field": "password"}
      ]
    }
  ]
}
```

File: api/reference/configuration_script_payloads.md (hunk 8 above)
Comment (agrare) [reply api/reference/configuration_script_payloads.md:67]:
```
POST /api/configuration_script_payloads/:id/authentications 
{
  "action": "create",
  "description": "Description",
  "name": "my credential",
  "type": "ManageIQ::Providers::AnsibleTower::AutomationManager::Credential"
}
```

File: api/reference/configuration_script_payloads.md (hunk 5 above)
Comment (rwellon) [reply api/reference/configuration_script_payloads.md:118]: These examples are added.

File: api/reference/configuration_script_payloads.md (hunk 6 above)
Comment (rwellon) [reply api/reference/configuration_script_payloads.md:202]: This example is added.

File: api/reference/configuration_script_payloads.md (hunk 7 above)
Comment (rwellon) [reply api/reference/configuration_script_payloads.md:254]: This example is added.

File: api/reference/configuration_script_payloads.md (hunk 3 above)
Comment (rwellon) [reply api/reference/configuration_script_payloads.md:309]: This example is added.

File: api/reference/configuration_script_payloads.md (hunk 4 above)
Comment (rwellon) [reply api/reference/configuration_script_payloads.md:341]: This example is added.

File: api/reference/configuration_script_payloads.md [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+``` json
+{
+  "action": "edit",
+  "resources": [
+    {
+      "id": 1,
+      "credentials": [
+        {"my-cred-user": {"credential_ref": "my-credential", "credential_field": "userid"},
+        {"my-cred-password": {"credential_ref": "my-credential", "credential_field": "password"}
```
Comment (agrare) [inline api/reference/configuration_script_payloads.md:189]: Looks like a typo in the suggestion :wink: 

```suggestion
        {"my-cred-user": {"credential_ref": "my-credential", "credential_field": "userid"}},
        {"my-cred-password": {"credential_ref": "my-credential", "credential_field": "password"}}
```

Comment (rwellon) [reply api/reference/configuration_script_payloads.md]: I have fixed this typo, Thanks!

**Review by agrare — APPROVED** [review]

**Comment by Fryguy** [comment]
Backported to `quinteros` in commit e3b8acb8d234eb9a1dafd60056e346a58dcbe05c.

```text
commit e3b8acb8d234eb9a1dafd60056e346a58dcbe05c
Author: Adam Grare <adam@grare.com>
Date:   Wed Dec 6 10:35:58 2023 -0500

    Merge pull request #1757 from rwellon/RW-IAE-API
    
    Create configuration_script_payloads API doc
    
    (cherry picked from commit 87501e16c4bb7d883de0d92624a4855b71acd6a3)
```

**Pull request merged**